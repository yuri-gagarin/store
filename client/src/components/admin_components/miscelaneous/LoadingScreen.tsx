import React from 'react'
import { Dimmer, Loader, Segment } from 'semantic-ui-react'

const LoaderExampleText: React.FC<{}> = (props): JSX.Element => {
  return (
    <div style={{ height: "100vh" }}> 
      <Segment style={{ height: "100vh" }}>
        <Dimmer active inverted>
          <Loader>Please Wait</Loader>
        </Dimmer>
      </Segment>
    </div>
  )
};

export default LoaderExampleText