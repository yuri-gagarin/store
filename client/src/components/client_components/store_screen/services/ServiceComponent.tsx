import React, { useState, useEffect } from "react";
import { Image, Icon } from "semantic-ui-react";
// css imports //
import "./css/storeServices.css";

interface Props {
  coverUrl?: string;
  pictureUrl?: string;
  description?: string;
};

const ServiceComponent: React.FC<Props> = ({ coverUrl, pictureUrl, description }): JSX.Element => {
  const [ covUrl, setCoverUrl ] = useState<string>("");
  const [ picUrl, setPicUrl ] = useState<string>("");

  const pullCoverUrl = (coverUrl: string | undefined) => {
    if (coverUrl) {
      setCoverUrl(coverUrl);
    } else {
      setCoverUrl("/images/services/service2.jpeg");
    }
  };
  const pullPicUrl = (pictureUrl: string | undefined) => {
    if (pictureUrl) {
      setPicUrl(pictureUrl);
    } else {
      setPicUrl("/images/services/service1.jpeg");
    }
  };
  // set image url or default image //
  useEffect(() => {
    pullPicUrl(coverUrl);
    pullCoverUrl(pictureUrl);
  }, [coverUrl, pictureUrl]);
  return (
    <div className="storeServiceHolder">
      <div className="storeServiceHiddenImg">
        <Image src={picUrl} size="medium"  />
      </div>
      <div className="storeServiceCoverImg" >
        <Image src={covUrl} size="medium" />
      </div>
      <Icon name="plus"></Icon>

    </div>

    
  );
};

export default ServiceComponent;