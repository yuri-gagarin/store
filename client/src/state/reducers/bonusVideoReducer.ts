// types //
export const emptyBonusVideoData = (): IBonusVideoData => {
  return {
    _id: "",
    description: "",
    youTubeURL: "",
    vimeoURL: "",
    createdAt: ""
  }
};

export const initialBonusVideoState: IBonusVideoState = {
  loading: false,
  responseMsg: "",
  currentBonusVideoData: emptyBonusVideoData(),
  loadedBonusVideos: [],
  error: null
};

const serviceReducer = (state: IBonusVideoState = initialBonusVideoState, action: BonusVideoAction): IBonusVideoState => {
  switch (action.type) {
    case "GET_ALL_BONUS_VIDEOS": 
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        loadedBonusVideos: [ ...action.payload.loadedBonusVideos ],
        error: action.payload.error
      };
    case "GET_BONUS_VIDEO": 
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        currentBonusVideoData: { ...action.payload.currentBonusVideoData },
        error: action.payload.error
      };
    case "SET_CURRENT_BONUS_VIDEO": 
      return {
        ...state,
        currentBonusVideoData: { ...action.payload.currentBonusVideoData }
      };
    case "CLEAR_CURRENT_BONUS_VIDEO": 
      return {
        ...state,
        currentBonusVideoData: { ...emptyBonusVideoData() }
      };
    case "CREATE_BONUS_VIDEO":
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        currentBonusVideoData: { ...action.payload.newBonusVideo },
        loadedBonusVideos: [ ...state.loadedBonusVideos, action.payload.newBonusVideo ],
        error: action.payload.error
      };
    case "EDIT_BONUS_VIDEO": 
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        loadedBonusVideos: [ ...action.payload.loadedBonusVideos],
        error: action.payload.error
      };
    case "DELETE_BONUS_VIDEO": 
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        currentBonusVideoData: { ...emptyBonusVideoData() },
        loadedBonusVideos: [ ...action.payload.loadedBonusVideos ],
        error:action.payload.error
      };
    case "SET_BONUS_VIDEO_ERROR": 
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        error: action.payload.error
      };
    case "ClEAR_BONUS_VIDEO_ERROR": 
      return {
        ...state,
        loading:action.payload.loading,
        responseMsg: action.payload.responseMsg,
        error: action.payload.error
      };
    default: return state;
  }
};

export default serviceReducer;