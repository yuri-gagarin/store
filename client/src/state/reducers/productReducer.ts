// types //
export const emptyProductData = (): IProductData => {
  return {
    _id: "",
    name: "",
    details: "",
    description: "",
    price: "",
    images: [],
    createdAt: "",
  }
};

export const initialProductState: IProductState = {
  loading: false,
  responseMsg: "",
  currentProductData: emptyProductData(),
  loadedProducts: [],
  productFormOpen: false,
  error: null
};

const productReducer = (state: IProductState = initialProductState, action: ProductAction): IProductState => {
  switch (action.type) {
    case "GET_ALL_PRODUCTS": 
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        loadedProducts: [ ...action.payload.loadedProducts],
        error: action.payload.error
      };
    case "DISPATCH_PRODUCT_API_REQUEST": 
      return {
        ...state,
        loading: action.payload.loading,
        error: action.payload.error
      };
    case "GET_PRODUCT": 
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        currentProductData: { ...action.payload.currentProductData },
        error: action.payload.error
      };
    case "SET_CURRENT_PRODUCT": 
      return {
        ...state,
        currentProductData: { ...action.payload.currentProductData }
      };
    case "CLEAR_CURRENT_PRODUCT": 
      return {
        ...state,
        currentProductData: { ...emptyProductData() }
      };
    case "OPEN_PRODUCT_FORM": 
      return {
        ...state,
        productFormOpen: action.payload.productFormOpen
      };
    case "CLOSE_PRODUCT_FORM": 
      return {
        ...state,
        productFormOpen: action.payload.productFormOpen
      };
    case "UPLOAD_NEW_PRODUCT_IMG": 
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        currentProductData: { ...action.payload.editedProduct },
        loadedProducts: [ ...action.payload.loadedProducts ],
        error: action.payload.error
      };
    case "DELETE_PRODUCT_IMG": {
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        currentProductData: { ...action.payload.editedProduct },
        loadedProducts: [ ...action.payload.loadedProducts ],
        error: action.payload.error
      }
    }
    case "UPDATE_PRODUCT_IMGS": 
      return {
        ...state,
        currentProductData: { ...action.payload.editedProduct }
      };
    case "CREATE_PRODUCT":
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        currentProductData: { ...action.payload.newProduct },
        loadedProducts: [ ...state.loadedProducts, action.payload.newProduct ],
        error: action.payload.error
      };
    case "EDIT_PRODUCT": 
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        currentProductData: { ...action.payload.editedProduct },
        loadedProducts: [ ...action.payload.loadedProducts ],
        error: action.payload.error
      };
    case "DELETE_PRODUCT": 
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        currentProductData: { ...emptyProductData() },
        loadedProducts: [ ...action.payload.loadedProducts ],
        error:action.payload.error
      };
    case "SET_PRODUCT_ERROR": 
      return {
        ...state,
        loading: action.payload.loading,
        responseMsg: action.payload.responseMsg,
        error: action.payload.error
      };
    case "ClEAR_PRODUCT_ERROR": 
      return {
        ...state,
        loading:action.payload.loading,
        responseMsg: action.payload.responseMsg,
        error: action.payload.error
      };
    default: return state;
  }
};

export default productReducer;