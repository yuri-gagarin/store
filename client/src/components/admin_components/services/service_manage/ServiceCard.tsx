import React, { useState } from "react";
import { Button, Grid } from "semantic-ui-react";
// additional components //
import EditControls from "./EditControls"
// css imports //
import "./css/serviceCard.css";
// actions and state /
import { deleteService } from "../actions/APIServiceActions";
import { setCurrentService, clearCurrentService} from "../actions/UIServiceActions";
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// additional dependencies //
import { withRouter, RouteComponentProps } from "react-router-dom";
// helpers //
import { ConvertDate } from "../../../helpers/displayHelpers";

interface Props extends RouteComponentProps {
  service: IServiceData;
  imageCount: number;
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}

const ServiceCard: React.FC<Props> = ({ service, imageCount, state, dispatch, history }): JSX.Element => {
  const [ editing, setEditing ] = useState<boolean>(false);
  const baseUrl = "/admin/home/my_services/manage"
  const { _id, name, price, description, images, createdAt, editedAt } = service;

  const handleServiceOpen = (e: React.MouseEvent<HTMLButtonElement>): void => {
    history.push(baseUrl + "/view");
  }
  const handleServiceEdit = (e: React.MouseEvent<HTMLButtonElement>): void => {
    if (!editing) {
      setCurrentService(_id, dispatch, state);
      history.push(baseUrl + "/edit");
      setEditing(true);
    } else {
      clearCurrentService(dispatch);
      history.push(baseUrl);
      setEditing(false);
    }
  }
  const handleServiceDelete = (e: React.MouseEvent<HTMLButtonElement>): void => {
    deleteService( _id, dispatch, state)
      .then((success) => {
        
      })
  }

  return (
    <React.Fragment>
      <Grid.Row style={{ border: "3px solid green", padding: "0.5em", marginTop: "2em" }}>
        <Grid.Column computer={12} mobile={16} style={{ border: '2px solid red'}}>
          <h3>Name: {name}</h3>
          <h5>Description: {description}</h5>
          <h5>Number of Images: {imageCount}</h5>
          <p><span>Created At: </span>{ConvertDate.international(createdAt)}</p>
        </Grid.Column> 
        <Grid.Column computer={4} mobile={16} style={{ border: '2px solid blue'}}>
          <Button 
            inverted
            color="green"
            className="serviceCardBtn" 
            content="Open" 
            onClick={handleServiceOpen} 
          />
          <Button 
            inverted
            color="orange"
            className="serviceCardBtn" 
            content="Edit" 
            onClick={handleServiceEdit}  
          />
          <Button 
            inverted
            color="red"
            className="serviceCardBtn" 
            content="Delete" 
            onClick={handleServiceDelete}
          />
        </Grid.Column> 
      </Grid.Row>
      {  editing ?  <EditControls handleServiceEdit={handleServiceEdit}/>: null }
    </React.Fragment>
    
  );
};

export default withRouter(ServiceCard);