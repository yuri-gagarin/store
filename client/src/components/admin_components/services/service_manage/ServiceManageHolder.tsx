import React, { useContext, useEffect, useState } from "react";
import { Button, Grid } from "semantic-ui-react";
// additional components //
import ServiceFormHolder from "../forms/ServiceFormHolder";
import ServiceCard from "./ServiceCard";
import LoadingScreen from "../../miscelaneous/LoadingScreen";
// actions and state //
import { getAllServices } from "../actions/APIServiceActions";
import { IGlobalAppState, AppAction } from "../../../../state/Store";
// additional dependencies //
import { withRouter, RouteComponentProps, useRouteMatch, Route } from "react-router-dom";

interface Props extends RouteComponentProps {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
};

const ServiceManageHolder: React.FC<Props> = ({ state, dispatch, history }): JSX.Element => {
  const { loadedServices } = state.serviceState;
  // local state //
  const [ pageLoaded, setPageLoaded ] = useState<boolean>(false);
  const match = useRouteMatch("/admin/home/my_services/manage");

  const handleBack = () => {
    history.goBack();
  };
  useEffect(() => {
    getAllServices(dispatch)
      .then((success) => {
        if (success) {
          setPageLoaded(true);
        }
      });
  }, []); 

  if (pageLoaded) {
    return (
      <Grid padded stackable columns={2}>
        <Route path={match?.url + "/edit"}> 
          <Grid.Row>
            <Grid.Column computer={12} tablet={6} mobile={16}>
              <h3>Editing Service: { state.serviceState.currentServiceData.name }</h3>
              <Button inverted color="green" content="Back" onClick={handleBack}></Button>
            </Grid.Column>
          </Grid.Row>
          <ServiceFormHolder state={state} dispatch={dispatch} />
        </Route>
        <Route exact path={match?.url}>
          <Grid.Row>
            <Grid.Column computer={12} tablet={8} mobile={16}>
            {
              loadedServices.map((service) => {
                return (
                  <ServiceCard 
                    key={service._id}
                    service={service}
                    imageCount={service.images.length}
                    state={state}
                    dispatch={dispatch}
                  />
                );
              })
            }
            </Grid.Column>
            <Grid.Column computer={4} tablet={8} mobile={16}>
              
            </Grid.Column>
          </Grid.Row>
        </Route>
      </Grid>
    );
  } else {
    return (
      <LoadingScreen />
    );
  }
};

export default withRouter(ServiceManageHolder);