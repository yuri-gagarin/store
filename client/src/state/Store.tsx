import React, { createContext, useReducer } from "react";
import { indexReducer, rootState } from "./reducers/indexReducer";
import { initialStoreState, StoreAction, IStoreState } from "./reducers/storeReducer";
// global app state //
export type IGlobalAppState = {
  storeState: IStoreState;
}
// app actions //
export type AppAction = StoreAction;
// global app context //
export interface IGlobalAppContext {
  state: IGlobalAppState;
  dispatch: React.Dispatch<AppAction>;
}
export type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;
// context initialization //
const initialContext: IGlobalAppContext = {
  state: {
    storeState: { ...initialStoreState }
  },
  dispatch: (value: AppAction): void => {}
} 
const Store = createContext<IGlobalAppContext>(initialContext);

export const StateProvider: React.FC<{}> = ({ children }): JSX.Element => {
  const [ globalState, dispatch ] = useReducer<CombinedReducer>(indexReducer, rootState as IGlobalAppState);
  return  (
    <Store.Provider value={{ state: globalState, dispatch: dispatch }}>
      {children}
    </Store.Provider>
  );
};

