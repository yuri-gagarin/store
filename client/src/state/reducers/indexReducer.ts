import combineReducers from "react-combine-reducers";
import bonusVideoReducer, { initialBonusVideoState } from "./bonusVideoReducer";
import productReducer, { initialProductState } from "./productReducer";
import storeReducer, { initialStoreState } from "./storeReducer";
import storeItemReducer, { initialStoreItemState } from "./storeItemReducer";
import serviceReducer, { initialServiceState } from "./serviceReducer";
import { IGlobalAppState, AppAction } from "../Store";

type CombinedReducer = (state: IGlobalAppState, action: AppAction) => IGlobalAppState;

export const [indexReducer, rootState ] = combineReducers<CombinedReducer>({
  bonusVideoState: [bonusVideoReducer, initialBonusVideoState],
  productState: [productReducer, initialProductState],
  storeState: [storeReducer, initialStoreState],
  storeItemState: [storeItemReducer, initialStoreItemState],
  serviceState: [serviceReducer, initialServiceState]
});