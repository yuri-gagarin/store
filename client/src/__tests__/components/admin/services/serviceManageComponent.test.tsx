import React, { useReducer, useEffect } from "react";
import { mount, ReactWrapper, shallow } from "enzyme";
import moxios from "moxios";
import fetchMock from "jest-fetch-mock";
import { MemoryRouter as Router } from "react-router-dom";
// components //
import ServiceManageHolder from "../../../../components/admin_components/services/service_manage/ServiceManageHolder";
import LoadingScreen  from "../../../../components/admin_components/miscelaneous/LoadingScreen";
import TestRenderer from 'react-test-renderer';
import { act } from "react-dom/test-utils";
import { renderHook } from "@testing-library/react-hooks";
import { createMockServices } from "../../../../test_helpers/serviceHelpers";
import { generateCleanState } from "../../../../test_helpers/miscHelpers";
import { initialContext, StateProvider, IGlobalAppState } from "../../../../state/Store";
import ServiceCard from "../../../../components/admin_components/services/service_manage/ServiceCard";
import { indexReducer, rootState } from "../../../../state/reducers/indexReducer";
import { RenderResult } from "@testing-library/react";
import ServiceComponent from "../../../../components/client_components/store_screen/services/ServiceComponent";

describe("Service Manage Holder Tests", () => {
  beforeEach(() => {
    moxios.install();
  });
  afterEach(() => {
    moxios.uninstall();
  });
  describe("Default Component state at first render", () => {
    let component: ReactWrapper; let loadingScreen: ReactWrapper;
    let state: IGlobalAppState; 

    beforeAll(() => {
      state = generateCleanState();
      component = mount(
        <Router>
          <StateProvider>
            <ServiceManageHolder state={state} dispatch={jest.fn}/>
          </StateProvider>
        </Router>
      )
    });
    it("Should correctly render", () => {
      expect(component).toMatchSnapshot();
    });
    it("Should render a 'LoadingScreen' Component before an API call resolves", () => {
      loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1)
    });
  });

  describe("State after a successful API call", () => {
    let component: ReactWrapper; let state: IGlobalAppState;
    let services: IServiceData[];


    beforeAll( async () => {
      moxios.install();
      state = generateCleanState();
      services = createMockServices(5);

      component = mount(
        <Router>
          <StateProvider>
            <ServiceManageHolder state={state} dispatch={jest.fn} />

          </StateProvider>
        </Router>
      );
      await act( async () => {
        moxios.stubRequest('/api/services/', {
          status: 200,
          responseText: 'hello',
          response: {
            responseMsg: "All ok",
            services: services
          }
        });
      });
      
      /*
      await moxios.wait(jest.fn);
      await act(async() => {
        const request = moxios.requests.mostRecent();
        await request.respondWith({
          response: {
            responseMsg: "All ok",
            services: services
          }
        })
      });
      */
    });

    afterAll(() => {
      moxios.uninstall();
    })
    
    it('Should correctly render initial Loading Screen', () => {
      let loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(1);
    });

    it('Should NOT render initial Loading Screen after successful', () => {
      act(() => {
        component.update();
      });
      let loadingScreen = component.find(LoadingScreen);
      expect(loadingScreen.length).toEqual(0);
    });

    it("Should render correct ServiceManageHolder Component", () => {
      const serviceManageHolderComp = component.find("#serviceManageHolder");
      expect(serviceManageHolderComp.at(0)).toBeDefined();
      expect(serviceManageHolderComp.at(0)).toMatchSnapshot();
    });

    it("Should properly render all ServiceCard components", () => {
      const comp = shallow(<ServiceManageHolder state={generateCleanState()} dispatch={jest.fn} />)
      const globalState = generateCleanState();
      const newState: IGlobalAppState = {
        ...globalState,
        serviceState: {
          ...globalState.serviceState,
          responseMsg: "All Ok",
          loadedServices: [ ...services ]
        }
      }
      const updated = comp.setProps({ state: newState});
      console.log(updated.find(ServiceCard).length)
      /*
      const serviceCards = component.find(ServiceCard);
      expect(serviceCards.at(0)).toBeDefined();
      expect(serviceCards.length).toEqual(services.length);
      */
    })
  });
});

/*
    describe("State after a successful API call", () => {
      let component: ReactWrapper; let loadingScreen: ReactWrapper;
      let Services: IServiceData[];

      beforeEach(() => {
        moxios.install();
      });
      afterEach(() => {
        moxios.uninstall();
      })
      
      it("Should fetch the request", async () => {
        await act(async () => {
          await mount(
            <Router>
              <StateProvider>
                <ServiceManageHolder />

              </StateProvider>
            </Router>
          );
        });
        expect(fetchMock).toHaveBeenCalledWith("thisurl")
      });

      it('Should display error if api request fail', async () => {
      
        let comp: ReactWrapper;
        comp = mount(
          <Router>
            <StateProvider>
              <ServiceManageHolder />

            </StateProvider>
          </Router>
        );
        await moxios.wait(jest.fn);
        await act(async() => {
          const request = moxios.requests.mostRecent()
          await request.respondWith({
            response: {
              responseMsg: "All ok",
              Services: createMockServices(5)
            }
          })
        });
        act(() => {
          comp.update();
        })
        let ServiceCards = comp.find(ServiceCard)
        console.log(ServiceCards.length)
      });

      it('should display error if api request fail', async () => {
      
        let comp: ReactWrapper | null;
        comp = mount(
          <Router>
            <StateProvider>
              <ServiceManageHolder />

            </StateProvider>
          </Router>
        );
        await moxios.wait(jest.fn);
        await act(async() => {
          const request = moxios.requests.mostRecent()
          await request.reject(new Error("Boo"));
        });
        act(() => {
          comp!.update();
          console.log(comp?.props())
        })
    
      });
      /*
      it("Should NOT render the 'LoadingScreen' Component", () => {
        let loadingScreen = component.find(LoadingScreen);
        expect(loadingScreen.length).toEqual(0);
      });

      it("Should render the Service Manage grid", () => {
        //console.log(component.props())
      })
    });
    */