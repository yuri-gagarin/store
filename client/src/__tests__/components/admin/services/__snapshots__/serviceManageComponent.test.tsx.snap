// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Service Manage Holder Tests Default Component state at first render Should correctly render 1`] = `
<MemoryRouter>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "jox4jb",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "jox4jb",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Component>
      <withRouter(ServiceManageHolder)
        dispatch={[Function]}
        state={
          Object {
            "bonusVideoState": Object {
              "currentBonusVideoData": Object {
                "_id": "",
                "createdAt": "",
                "description": "",
                "vimeoURL": "",
                "youTubeURL": "",
              },
              "error": null,
              "loadedBonusVideos": Array [],
              "loading": false,
              "responseMsg": "",
            },
            "productState": Object {
              "currentProductData": Object {
                "_id": "",
                "createdAt": "",
                "description": "",
                "details": "",
                "images": Array [],
                "name": "",
                "price": "",
              },
              "error": null,
              "loadedProducts": Array [],
              "loading": false,
              "responseMsg": "",
            },
            "serviceState": Object {
              "currentServiceData": Object {
                "_id": "",
                "createdAt": "",
                "description": "",
                "images": Array [],
                "name": "",
                "price": "",
              },
              "error": null,
              "loadedServices": Array [],
              "loading": false,
              "responseMsg": "",
            },
            "storeItemState": Object {
              "currentStoreItemData": Object {
                "_id": "",
                "categories": Array [],
                "createdAt": "",
                "description": "",
                "details": "",
                "images": Array [],
                "name": "",
                "price": "",
                "storeId": "",
                "storeName": "",
              },
              "error": null,
              "loadedStoreItems": Array [],
              "loading": false,
              "numberOfItems": 0,
              "responseMsg": "",
            },
            "storeState": Object {
              "currentStoreData": Object {
                "_id": "",
                "createdAt": "",
                "description": "",
                "images": Array [],
                "title": "",
              },
              "error": null,
              "loadedStores": Array [],
              "loading": false,
              "responseMsg": "",
            },
          }
        }
      >
        <ServiceManageHolder
          dispatch={[Function]}
          history={
            Object {
              "action": "POP",
              "block": [Function],
              "canGo": [Function],
              "createHref": [Function],
              "entries": Array [
                Object {
                  "hash": "",
                  "key": "jox4jb",
                  "pathname": "/",
                  "search": "",
                  "state": undefined,
                },
              ],
              "go": [Function],
              "goBack": [Function],
              "goForward": [Function],
              "index": 0,
              "length": 1,
              "listen": [Function],
              "location": Object {
                "hash": "",
                "key": "jox4jb",
                "pathname": "/",
                "search": "",
                "state": undefined,
              },
              "push": [Function],
              "replace": [Function],
            }
          }
          location={
            Object {
              "hash": "",
              "key": "jox4jb",
              "pathname": "/",
              "search": "",
              "state": undefined,
            }
          }
          match={
            Object {
              "isExact": true,
              "params": Object {},
              "path": "/",
              "url": "/",
            }
          }
          state={
            Object {
              "bonusVideoState": Object {
                "currentBonusVideoData": Object {
                  "_id": "",
                  "createdAt": "",
                  "description": "",
                  "vimeoURL": "",
                  "youTubeURL": "",
                },
                "error": null,
                "loadedBonusVideos": Array [],
                "loading": false,
                "responseMsg": "",
              },
              "productState": Object {
                "currentProductData": Object {
                  "_id": "",
                  "createdAt": "",
                  "description": "",
                  "details": "",
                  "images": Array [],
                  "name": "",
                  "price": "",
                },
                "error": null,
                "loadedProducts": Array [],
                "loading": false,
                "responseMsg": "",
              },
              "serviceState": Object {
                "currentServiceData": Object {
                  "_id": "",
                  "createdAt": "",
                  "description": "",
                  "images": Array [],
                  "name": "",
                  "price": "",
                },
                "error": null,
                "loadedServices": Array [],
                "loading": false,
                "responseMsg": "",
              },
              "storeItemState": Object {
                "currentStoreItemData": Object {
                  "_id": "",
                  "categories": Array [],
                  "createdAt": "",
                  "description": "",
                  "details": "",
                  "images": Array [],
                  "name": "",
                  "price": "",
                  "storeId": "",
                  "storeName": "",
                },
                "error": null,
                "loadedStoreItems": Array [],
                "loading": false,
                "numberOfItems": 0,
                "responseMsg": "",
              },
              "storeState": Object {
                "currentStoreData": Object {
                  "_id": "",
                  "createdAt": "",
                  "description": "",
                  "images": Array [],
                  "title": "",
                },
                "error": null,
                "loadedStores": Array [],
                "loading": false,
                "responseMsg": "",
              },
            }
          }
        >
          <LoaderExampleText>
            <div
              style={
                Object {
                  "height": "100vh",
                }
              }
            >
              <Segment
                style={
                  Object {
                    "height": "100vh",
                  }
                }
              >
                <div
                  className="ui segment"
                  style={
                    Object {
                      "height": "100vh",
                    }
                  }
                >
                  <Dimmer
                    active={true}
                    inverted={true}
                  >
                    <DimmerInner
                      active={true}
                      inverted={true}
                    >
                      <Ref
                        innerRef={
                          Object {
                            "current": <div
                              class="ui active transition visible inverted dimmer"
                              style="display: flex;"
                            >
                              <div
                                class="content"
                              >
                                <div
                                  class="ui text loader"
                                >
                                  Please Wait
                                </div>
                              </div>
                            </div>,
                          }
                        }
                      >
                        <RefFindNode
                          innerRef={
                            Object {
                              "current": <div
                                class="ui active transition visible inverted dimmer"
                                style="display: flex;"
                              >
                                <div
                                  class="content"
                                >
                                  <div
                                    class="ui text loader"
                                  >
                                    Please Wait
                                  </div>
                                </div>
                              </div>,
                            }
                          }
                        >
                          <div
                            className="ui active transition visible inverted dimmer"
                            onClick={[Function]}
                          >
                            <div
                              className="content"
                            >
                              <Loader>
                                <div
                                  className="ui text loader"
                                >
                                  Please Wait
                                </div>
                              </Loader>
                            </div>
                          </div>
                        </RefFindNode>
                      </Ref>
                    </DimmerInner>
                  </Dimmer>
                </div>
              </Segment>
            </div>
          </LoaderExampleText>
        </ServiceManageHolder>
      </withRouter(ServiceManageHolder)>
    </Component>
  </Router>
</MemoryRouter>
`;

exports[`Service Manage Holder Tests State after a successful API call Should render correct ServiceManageHolder Component 1`] = `
<Grid
  columns={2}
  id="serviceManageHolder"
  padded={true}
  stackable={true}
>
  <div
    className="ui stackable padded two column grid"
    id="serviceManageHolder"
  >
    <Route
      path="undefined/edit"
    />
    <Route
      exact={true}
    >
      <GridRow>
        <div
          className="row"
        >
          <GridColumn
            computer={12}
            mobile={16}
            tablet={8}
          >
            <div
              className="twelve wide computer sixteen wide mobile eight wide tablet column"
            />
          </GridColumn>
          <GridColumn
            computer={4}
            mobile={16}
            tablet={8}
          >
            <div
              className="four wide computer sixteen wide mobile eight wide tablet column"
            />
          </GridColumn>
        </div>
      </GridRow>
    </Route>
  </div>
</Grid>
`;
